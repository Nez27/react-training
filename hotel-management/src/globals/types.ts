type TUser = {
  id: number;
  name: string;
  identifiedCode: string;
  address: string;
  phone: string;
  roomId: number;
};

type TRoom = {
  id: number;
  name: string;
  amount: number;
  price: number;
  discount: number;
  description: string;
  status: boolean;
};

type TStateSchema = {
  [key: string]: {
    value?: string;
    error?: string;
  };
};

type TKeyValue = {
  [key: string]: boolean | undefined | string;
};

type TKeyString = {
  [key: string]: string;
};

type TPropValues = 'value' | 'error' | boolean;

type TValidator = {
  [key: string]: {
    required?: boolean;
    validator?: {
      func?: (value: string) => boolean;
      error?: string;
    };
  };
};

type TResponse = {
  statusCode: number;
  msg: string;
};

export type {
  TUser,
  TRoom,
  TStateSchema,
  TKeyValue,
  TKeyString,
  TPropValues,
  TValidator,
  TResponse,
};
