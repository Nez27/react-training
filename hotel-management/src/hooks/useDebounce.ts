import { useEffect, useState } from 'react';

/**
 * The debounce hooks use to execute the callback function every x delay time
 * @param value The value need to be pass
 * @param delay The number of time want to delay
 * @returns Return the value after delay number
 */
const useDebounce = <T>(value: T, delay: number): T => {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const timer = setTimeout(() => setDebouncedValue(value), delay);

    return () => {
      clearTimeout(timer);
    };
  }, [value, delay]);

  return debouncedValue;
};

export { useDebounce };
