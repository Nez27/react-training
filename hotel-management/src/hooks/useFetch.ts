import { useEffect, useState } from 'react';

// Constants
import { BASE_URL } from '../constants/path';
import { searchQuery } from '../helpers/utils';
import { DEFAULT_ORDER_BY, DEFAULT_SORT_BY } from '../constants/config';

export const useFetch = (
  path: string,
  phoneNum: string,
  tempSortBy: string,
  tempOrderBy: string,
  reload?: boolean,
) => {
  const [data, setData] = useState(null);
  const [isPending, setIsPending] = useState(false);
  const [errorMsg, setErrorMsg] = useState<string | null>(null);

  useEffect(() => {
    // Clear data
    setData(null);

    const fetchData = async () => {
      setIsPending(true);

      // Set default value
      const sortBy = tempSortBy ? tempSortBy : DEFAULT_SORT_BY;
      const orderBy = tempOrderBy ? tempOrderBy : DEFAULT_ORDER_BY;

      // Query search
      const query = searchQuery(phoneNum, sortBy, orderBy);

      try {
        const response = await fetch(BASE_URL + path + '?' + query);
        const json = await response.json();

        if (!response.ok)
          throw new Error(
            `Error code: ${response.status} \n Messages: ${response.statusText}`,
          );

        setIsPending(false);
        setData(json);
        setErrorMsg(null);
      } catch (error) {
        setErrorMsg(`Could not fetch data.\n ${error}`);
        setIsPending(false);
      }
    };

    fetchData();
  }, [path, reload, phoneNum, tempOrderBy, tempSortBy]);
  return { data, isPending, errorMsg };
};
