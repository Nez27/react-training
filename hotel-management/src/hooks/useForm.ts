import { useState, useEffect, useCallback, ChangeEvent } from 'react';

// Utils
import {
  ERROR,
  VALUE,
  getPropValues,
  isObject,
  isRequired,
} from '../helpers/utils';

// Types
import { TKeyValue, TValidator } from '../globals/types';

/**
 * Custom hooks to validate your Form...
 *
 * @param stateSchema model you stateSchema.
 * @param stateValidatorSchema model your validation.
 * @param submitFormCallback function to be execute during form submission.
 * @returns
 */
const useForm = (
  stateSchema = {},
  stateValidatorSchema = {} as TValidator,
  submitFormCallback: (values: TKeyValue) => void,
  initialValue: string = '',
) => {
  const [values, setValues] = useState(getPropValues(stateSchema, VALUE));
  const [errors, setErrors] = useState(getPropValues(stateSchema, ERROR));
  const [dirty, setDirty] = useState(getPropValues(stateSchema));
  const [disable, setDisable] = useState(true);
  const [isDirty, setIsDirty] = useState(false);

  // Get a local copy of stateSchema
  useEffect(() => {
    setInitialErrorState(initialValue);
    setDisable(true);

    // If initial value true, setValues again from stateSchema
    // and enabled button
    if (initialValue) {
      setValues({});
      setValues(getPropValues(stateSchema, VALUE));
      setDisable(false);
    }
  }, [initialValue]); // eslint-disable-line

  // Validate fields in forms
  const validateFormFields = useCallback(
    (name: string, value: string) => {
      const validator = stateValidatorSchema;
      // Making sure that stateValidatorSchema name is same in
      // stateSchema
      if (!validator[name]) return;

      const field = validator[name];

      let error = '';

      // Skip check id field
      if (name !== 'id') {
        error = isRequired(value, field!.required);

        if (isObject(field['validator']) && error === '') {
          const fieldValidator = field['validator'];

          // Test the function callback if the value is meet the criteria
          const testFunc = fieldValidator!['func'];
          if (!testFunc!(value)) {
            error = fieldValidator!['error']!;
          }
        }
      }

      return error;
    },
    [stateValidatorSchema],
  );

  // Set Initial Error State
  // When hooks was first rendered...
  const setInitialErrorState = useCallback(
    (initialValue: string) => {
      Object.keys(errors).map((name) =>
        setErrors((prevState) => ({
          ...prevState,
          [name]: !initialValue // Skip error when initialValue have values
            ? validateFormFields(name, values[name] as string)
            : '',
        })),
      );
    },
    [errors, values, validateFormFields],
  );

  // Used to disable submit button if there's a value in errors
  // or the required field in state has no value.
  // Wrapped in useCallback to cached the function to avoid intensive memory leaked
  // in every re-render in component
  const validateErrorState = useCallback(
    () => Object.values(errors).some((error) => error),
    [errors],
  );

  // For every changed in our state this will be fired
  // To be able to disable the button
  useEffect(() => {
    if (isDirty) {
      setDisable(validateErrorState());
    }
  }, [errors, isDirty, validateErrorState]);

  // Event handler for handling changes in input.
  const handleOnChange = useCallback(
    (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      setIsDirty(true);

      const name = (event.target! as HTMLInputElement).name;
      const value = (event.target! as HTMLInputElement).value;

      const error = validateFormFields(name, value);

      setValues((prevState) => ({ ...prevState, [name]: value }));
      setErrors((prevState) => ({ ...prevState, [name]: error }));
      setDirty((prevState) => ({ ...prevState, [name]: true }));
    },
    [validateFormFields],
  );

  const handleOnSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      // Making sure that there's no error in the state
      // before calling the submit callback function
      // and disabled button

      if (!validateErrorState()) {
        submitFormCallback(values);
        setDisable(true);
      }
    },
    [validateErrorState, submitFormCallback, values],
  );

  return {
    handleOnChange,
    handleOnSubmit,
    values,
    errors,
    disable,
    setValues,
    setErrors,
    dirty,
  };
};

export default useForm;
