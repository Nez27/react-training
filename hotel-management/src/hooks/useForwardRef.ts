import { ForwardedRef, useEffect, useRef } from 'react';

// Types
import { Nullable } from '@type/common';

const useForwardRef = <T>(
  ref: ForwardedRef<T>,
  initialValue: Nullable<T> = null
) => {
  const targetRef = useRef<T>(initialValue);

  useEffect(() => {
    if (!ref) return;

    if (typeof ref === 'function') {
      ref(targetRef.current);
    } else {
      ref.current = targetRef.current;
    }
  }, [ref]);

  return targetRef;
};

export { useForwardRef };
