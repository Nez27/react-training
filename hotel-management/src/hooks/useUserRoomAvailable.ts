import { useContext } from 'react';

// Contexts
import { UserRoomAvailableContext } from '@src/contexts/UserRoomAvailableContext';

/**
 * The hooks check the place call is in UserRoomAvailableProvider or not
 * @returns The context of UserRoomAvailableContext
 */
const useUserRoomAvailable = () => {
  const context = useContext(UserRoomAvailableContext);

  if (context === undefined) {
    throw new Error(
      'UserRoomAvailableContext was used outside of UserRoomAvailableProvider'
    );
  }

  return context;
};

export { useUserRoomAvailable };
