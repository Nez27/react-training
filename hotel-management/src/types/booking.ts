interface IBooking {
  id: number;
  startDate: string;
  endDate: string;
  status: boolean;
  amount: number;
  roomId: number;
  userId: number;
}

interface TBookingResponse {
  id: number;
  startDate: string;
  endDate: string;
  status: boolean;
  amount: number;
  rooms: {
    id: number
    name: string;
  } | null;
  users: {
    id: number;
    name: string;
  } | null;
}

export type { IBooking, TBookingResponse };
