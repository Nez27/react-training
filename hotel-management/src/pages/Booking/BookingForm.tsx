import { useCallback, useMemo } from 'react';
import { FormProvider, useForm } from 'react-hook-form';

// Styled
import Input from '@commonStyle/Input.ts';

// Constants
import { REQUIRED_FIELD_ERROR } from '@constant/formValidateMessage.ts';

// Components
import Form from '@component/Form/index.tsx';
import Select, { ISelectOptions } from '@component/Select/index.tsx';

// Hooks
import { useCreateBooking } from '@hook/bookings/useCreateBooking.ts';
import { useUpdateBooking } from '@hook/bookings/useUpdateBooking.ts';
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable.ts';

// Helpers
import {
  convertCurrencyToNumber,
  formatCurrency,
  getDayDiff,
} from '@helper/helper.ts';

// Services
import { getRoomById, updateRoomStatus } from '@service/roomServices.ts';
import { updateUserBookedStatus } from '@service/userServices.ts';

// Types
import { IBooking, TBookingResponse } from '@type/booking.ts';

interface IBookingFormProp {
  onCloseModal?: () => void;
  booking?: TBookingResponse;
}

interface IBookingForm extends Omit<IBooking, 'amount'> {
  amount: string;
}

const BookingForm = ({ onCloseModal, booking }: IBookingFormProp) => {
  const { roomsAvailable, usersAvailable, dispatch } = useUserRoomAvailable();
  const { isCreating, createBooking } = useCreateBooking();
  const { isUpdating, updateBooking } = useUpdateBooking();
  const isLoading = isCreating || isUpdating;
  const { id: editId, ...editValues } = { ...booking };
  const formMethods = useForm<IBookingForm>({
    defaultValues: editId
      ? {
          startDate: editValues.startDate,
          endDate: editValues.endDate,
          roomId: editValues.rooms?.id,
          userId: editValues.users?.id,
          amount: formatCurrency(editValues.amount!),
        }
      : {},
  });
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
    getValues,
    setValue,
  } = formMethods;

  // Submit form
  const onSubmit = async (newBooking: IBookingForm) => {
    if (!editId) {
      // Add request
      createBooking(
        { ...newBooking, amount: convertCurrencyToNumber(newBooking.amount) },
        {
          onSuccess: async () => {
            reset();
            onCloseModal?.();

            // Update room, user available in global state
            dispatch!({
              type: 'updateStatusUser',
              payload: [{ id: newBooking.userId, isBooked: true }],
            });
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: newBooking.roomId, status: true }],
            });

            // Update room, user available in server
            await updateRoomStatus(newBooking.roomId, true);
            await updateUserBookedStatus(newBooking.userId, true);
          },
        }
      );
    } else {
      // Edit request
      newBooking.id = editId!;
      updateBooking(
        { ...newBooking, amount: convertCurrencyToNumber(newBooking.amount) },
        {
          onSuccess: async () => {
            reset();
            onCloseModal?.();

            // Update room status in global state
            // Reset status old room
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: booking!.rooms!.id, status: false }],
            });

            // Change status new room
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: newBooking.roomId, status: true }],
            });

            // Update room status in server
            // Reset status old room
            await updateRoomStatus(booking!.rooms!.id, false);

            // Change status new room
            await updateRoomStatus(newBooking.roomId, true);
          },
        }
      );
    }
  };

  // Init user, room available to options choice
  const userOptions = useMemo(() => {
    const options: ISelectOptions[] = [];

    usersAvailable?.forEach((item) => {
      if (booking?.users?.id === item.id || !item.isBooked) {
        options.push({
          label: item.name!,
          value: item.id.toString(),
        });
      }
    });

    return options;
  }, [usersAvailable, booking?.users?.id]);

  const roomOptions = useMemo(() => {
    const options: ISelectOptions[] = [];

    roomsAvailable?.forEach((item) => {
      if (booking?.rooms?.id === item.id || !item.status) {
        options.push({
          label: item.name!,
          value: item.id.toString(),
        });
      }
    });

    return options;
  }, [roomsAvailable, booking?.rooms?.id]);

  // Calculate the final price
  const computePrice = useCallback(async () => {
    setValue('amount', 'Loading...');
    const startDateValue = getValues('startDate');
    const endDateValue = getValues('endDate');
    const roomValue = getValues('roomId');

    if (startDateValue && endDateValue && roomValue) {
      const daysDiff = getDayDiff(
        new Date(startDateValue),
        new Date(endDateValue)
      );

      // Fetch room data
      const room = await getRoomById(roomValue.toString());
      const amount = daysDiff * room.price;

      setValue('amount', formatCurrency(amount), { shouldValidate: true });
    }
  }, [getValues, setValue]);

  const startDateValidate = useMemo(
    () => new Date().toISOString().split('T')[0],
    []
  );

  const endDateValidate = () => {
    if (getValues('startDate')) {
      const date = new Date(getValues('startDate'));
      date.setDate(date.getDate() + 1);

      return date.toISOString().split('T')[0];
    }

    return;
  };

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Row label="User" error={errors?.userId?.message}>
          {userOptions.length ? (
            <Select
              ariaLabel="user-select"
              options={userOptions}
              id="userId"
              disable={Boolean(booking?.users)}
              optionsConfigForm={{
                valueAsNumber: true,
                onChange: () => trigger('userId'),
              }}
            />
          ) : (
            <p>No user available</p>
          )}
        </Form.Row>

        <Form.Row label="Room" error={errors?.roomId?.message}>
          {roomOptions.length ? (
            <Select
              ariaLabel="user-select"
              options={roomOptions}
              id="roomId"
              optionsConfigForm={{
                valueAsNumber: true,
                onChange: () => trigger('roomId'),
              }}
            />
          ) : (
            <p>No room available</p>
          )}
        </Form.Row>

        <Form.Row label="Start Date" error={errors?.startDate?.message}>
          <Input
            type="date"
            id="startDate"
            min={startDateValidate}
            {...register('startDate', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => {
                trigger('startDate');
                computePrice();
              },
            })}
          />
        </Form.Row>

        <Form.Row label="End Date" error={errors?.startDate?.message}>
          <Input
            type="date"
            id="endDate"
            min={endDateValidate()}
            disabled={!getValues('startDate')}
            {...register('endDate', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => {
                trigger('endDate');
                computePrice();
              },
            })}
          />
        </Form.Row>

        <Form.Row label="Amount">
          <Input
            type="text"
            id="amount"
            {...register('amount', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => trigger('amount'),
            })}
            readOnly
          />
        </Form.Row>

        <Form.Action>
          <Form.Button
            type="submit"
            name="submit"
            disabled={!isDirty || !isValid || isLoading}
          >
            {
              !editId 
                ? 'Add' 
                : 'Save'
            }
          </Form.Button>
          <Form.Button type="button" variations="secondary" onClick={onCloseModal}>
            Close
          </Form.Button>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default BookingForm;
