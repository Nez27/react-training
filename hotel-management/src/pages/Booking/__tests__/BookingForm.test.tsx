import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';
import {
  RenderResult,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react';

// Types
import { TBookingResponse } from '@src/types/booking';

// Components
import BookingForm from '../BookingForm';

// Hooks
import { useUserRoomAvailable } from '@src/hooks/useUserRoomAvailable';

const mockUpdateBooking = jest.fn();

jest.mock('@src/hooks/bookings/useUpdateBooking', () => ({
  ...jest.requireActual('@src/hooks/bookings/useUpdateBooking'),
  useUpdateBooking: jest.fn(() => ({
    isUpdating: false,
    updateBooking: mockUpdateBooking,
  })),
}));

jest.mock('@src/hooks/useUserRoomAvailable.ts');

const mockUseUserRoomAvailable = useUserRoomAvailable as jest.MockedFunction<
  typeof useUserRoomAvailable
>;

mockUseUserRoomAvailable.mockReturnValue({
  roomsAvailable: [
    { id: 1, name: 'Room 1', status: false },
    { id: 2, name: 'Room 2', status: false },
  ],
  usersAvailable: [
    { id: 1, name: 'User 1', isBooked: false },
    { id: 2, name: 'User 2', isBooked: false },
  ],
  dispatch: jest.fn(),
});

let result: {
  name: string;
  price: number;
};

describe('BookingForm', () => {
  const queryClient = new QueryClient();

  let wrapper: RenderResult | null = null;

  test('Should render correctly', () => {
    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <BookingForm />
        </BrowserRouter>
      </QueryClientProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('Should edit correctly', async () => {
    const tempBooking: TBookingResponse = {
      id: 1,
      startDate: '2024-11-21',
      endDate: '2024-11-22',
      amount: 1200,
      status: true,
      rooms: {
        id: 1,
        name: 'Room 1',
      },
      users: {
        id: 2,
        name: 'Room 2',
      },
    };

    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <BookingForm booking={tempBooking} />
        </BrowserRouter>
      </QueryClientProvider>
    );

    const submitBtn = wrapper!.getByText('Save').closest('button');

    mockUpdateBooking.mockImplementationOnce((newBooking, { onSuccess }) => {
      result = newBooking;
      onSuccess();
    });

    fireEvent.submit(submitBtn!);

    await waitFor(() => {
      expect(mockUpdateBooking).toHaveBeenCalled();
      expect(result).toEqual({
        amount: 1200,
        endDate: '2024-11-22',
        id: 1,
        roomId: 1,
        startDate: '2024-11-21',
        userId: 2,
      });
    });
  });
});
