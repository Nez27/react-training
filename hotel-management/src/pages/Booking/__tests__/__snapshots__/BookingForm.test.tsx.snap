// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-dlniIP bCPNqw"
      >
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            User:
          </label>
          <div>
            <select
              aria-label="user-select"
              class="sc-fujznN rEQue"
              id="userId"
              name="userId"
            >
              <option
                value="1"
              >
                User 1
              </option>
              <option
                value="2"
              >
                User 2
              </option>
            </select>
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Room:
          </label>
          <div>
            <select
              aria-label="user-select"
              class="sc-fujznN rEQue"
              id="roomId"
              name="roomId"
            >
              <option
                value="1"
              >
                Room 1
              </option>
              <option
                value="2"
              >
                Room 2
              </option>
            </select>
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Start Date:
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="startDate"
              min="2023-12-06"
              name="startDate"
              type="date"
            />
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            End Date:
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              disabled=""
              id="endDate"
              name="endDate"
              type="date"
            />
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Amount:
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="amount"
              name="amount"
              type="hidden"
            />
            <span>
              $0.00
            </span>
          </div>
        </div>
        <div
          class="sc-hKFymg dteuGK"
        >
          <button
            class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
            disabled=""
            name="submit"
            type="submit"
            variations="primary"
          >
            Add
          </button>
          <button
            class="sc-gtsqUy sc-iCoHzw hesuRJ cwhbgM"
            type="button"
            variations="secondary"
          >
            Close
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-dlniIP bCPNqw"
    >
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          User:
        </label>
        <div>
          <select
            aria-label="user-select"
            class="sc-fujznN rEQue"
            id="userId"
            name="userId"
          >
            <option
              value="1"
            >
              User 1
            </option>
            <option
              value="2"
            >
              User 2
            </option>
          </select>
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Room:
        </label>
        <div>
          <select
            aria-label="user-select"
            class="sc-fujznN rEQue"
            id="roomId"
            name="roomId"
          >
            <option
              value="1"
            >
              Room 1
            </option>
            <option
              value="2"
            >
              Room 2
            </option>
          </select>
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Start Date:
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="startDate"
            min="2023-12-06"
            name="startDate"
            type="date"
          />
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          End Date:
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            disabled=""
            id="endDate"
            name="endDate"
            type="date"
          />
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Amount:
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="amount"
            name="amount"
            type="hidden"
          />
          <span>
            $0.00
          </span>
        </div>
      </div>
      <div
        class="sc-hKFymg dteuGK"
      >
        <button
          class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
          disabled=""
          name="submit"
          type="submit"
          variations="primary"
        >
          Add
        </button>
        <button
          class="sc-gtsqUy sc-iCoHzw hesuRJ cwhbgM"
          type="button"
          variations="secondary"
        >
          Close
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
